nameko/containers.py

```diff
@@ -50,72 +50,59 @@ def get_container_cls(config):
     return import_from_path(class_path) or ServiceContainer
 
 
-def get_worker_ctx_cls(config):
-    class_path = config.get('WORKER_CTX_CLS')
-    return import_from_path(class_path) or WorkerContext
-
-
 def new_call_id():
     return str(uuid.uuid4())
 
 
-class WorkerContextBase(object):
-    """ Abstract base class for a WorkerContext
-    """
-    __metaclass__ = ABCMeta
+class WorkerContext(object):
 
-    def __init__(self, container, service, entrypoint, args=None, kwargs=None,
-                 data=None):
-        self.container = container
-        self.config = container.config  # TODO: remove?
+    _call_id = None
+    _call_id_stack = None
+    _parent_call_id_stack = None
 
-        self.parent_calls_tracked = self.config.get(
-            PARENT_CALLS_CONFIG_KEY, DEFAULT_PARENT_CALLS_TRACKED)
+    def __init__(
+        self, container, service, entrypoint, args=None, kwargs=None, data=None
+    ):
+        self.container = container
+        self.config = self.container.config
 
         self.service = service
         self.entrypoint = entrypoint
         self.service_name = self.container.service_name
 
         self.args = args if args is not None else ()
         self.kwargs = kwargs if kwargs is not None else {}
-
         self.data = data if data is not None else {}
 
-        self.parent_call_stack, self.unique_id = self._init_call_id()
-        self.call_id = '{}.{}.{}'.format(
-            self.service_name, self.entrypoint.method_name, self.unique_id
+        self._parent_call_id_stack = self.data.pop(
+            CALL_ID_STACK_CONTEXT_KEY, []
         )
-        n = -self.parent_calls_tracked
-        self.call_id_stack = self.parent_call_stack[n:]
-        self.call_id_stack.append(self.call_id)
-        try:
-            self.immediate_parent_call_id = self.parent_call_stack[-1]
-        except IndexError:
-            self.immediate_parent_call_id = None
 
-    @abstractproperty
-    def context_keys(self):
-        """ Return a tuple of keys describing data kept on this WorkerContext.
-        """
+    @property
+    def call_id_stack(self):
+        if self._call_id_stack is None:
+            parent_calls_tracked = self.container.config.get(
+                PARENT_CALLS_CONFIG_KEY, DEFAULT_PARENT_CALLS_TRACKED
+            )
+            stack_length = parent_calls_tracked + 1
+
```

Comment:
ok. don't feel that strongly. noted it because a) you only interact with the deque once (here) and b) you leave the attribute as a deque but convert it to a list every time it's read